client/client-multi/client-java-core/src/main/java/org/kaaproject/kaa/client/channel/failover/strategies/DefaultFailoverStrategy.java

```diff
@@ -0,0 +1,100 @@
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package org.kaaproject.kaa.client.channel.failover.strategies;
+
+import org.kaaproject.kaa.client.channel.TransportConnectionInfo;
+import org.kaaproject.kaa.client.channel.failover.FailoverDecision;
+import org.kaaproject.kaa.client.channel.failover.FailoverDecision.FailoverAction;
+import org.kaaproject.kaa.client.channel.failover.FailoverStatus;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Reference implementation for {@link FailoverStrategy}.
+ */
+public class DefaultFailoverStrategy implements FailoverStrategy {
+
+    private static final Logger LOG = LoggerFactory.getLogger(DefaultFailoverStrategy.class);
+
+    private static final long DEFAULT_BOOTSTRAP_SERVERS_RETRY_PERIOD = 2;
+    private static final long DEFAULT_OPERATION_SERVERS_RETRY_PERIOD = 2;
+    private static final long DEFAULT_NO_CONNECTIVITY_RETRY_PERIOD = 5;
+    private static final TimeUnit DEFAULT_TIME_UNIT = TimeUnit.SECONDS;
+
+    private long bootstrapServersRetryPeriod;
+    private long operationsServersRetryPeriod;
+    private long noConnectivityRetryPeriod;
+    private TimeUnit timeUnit;
+
+    public DefaultFailoverStrategy() {
+        this(DEFAULT_BOOTSTRAP_SERVERS_RETRY_PERIOD,
+                DEFAULT_OPERATION_SERVERS_RETRY_PERIOD,
+                DEFAULT_NO_CONNECTIVITY_RETRY_PERIOD,
+                DEFAULT_TIME_UNIT);
+    }
+
+    public DefaultFailoverStrategy(long bootstrapServersRetryPeriod,
+                                   long operationsServersRetryPeriod,
+                                   long noConnectivityRetryPeriod,
+                                   TimeUnit timeUnit) {
+        this.bootstrapServersRetryPeriod = bootstrapServersRetryPeriod;
+        this.operationsServersRetryPeriod = operationsServersRetryPeriod;
+        this.noConnectivityRetryPeriod = noConnectivityRetryPeriod;
+        this.timeUnit = timeUnit;
+    }
+
+    @Override
+    public FailoverDecision onFailover(FailoverStatus failoverStatus) {
+        LOG.trace("Producing failover decision for failover status: {}", failoverStatus);
+        switch (failoverStatus) {
+            case BOOTSTRAP_SERVERS_NA:
+                return new FailoverDecision(FailoverAction.RETRY, bootstrapServersRetryPeriod, timeUnit);
+            case CURRENT_BOOTSTRAP_SERVER_NA:
+                return new FailoverDecision(FailoverAction.USE_NEXT_BOOTSTRAP, bootstrapServersRetryPeriod, timeUnit);
+            case NO_OPERATION_SERVERS_RECEIVED:
+                return new FailoverDecision(FailoverAction.USE_NEXT_BOOTSTRAP, bootstrapServersRetryPeriod, timeUnit);
```

Comment:
As I understood, we postpone switching to the next bootstrap server. I thing we can send a request immediately.